### ISSUE TITLE
Iterating over BCOO arrays produces invalid indices.

### ISSUE THREAD

**Issue by tomsch420** [jax-ml/jax#24138]
### Description

When iterating a sparse tensor indices are produced that areinvalid for elements that did not exist in the BCOO.

Example to reproduce this bug:

```
from jax.experimental.sparse import BCOO
import jax.numpy as jnp

indices = jnp.array([[0, 0],
                     [0, 1],
                     [1, 0],
                     [2, 1],
                     [3, 0],
                     [3, 1]])
values = jnp.array([0, 0, 0, 0, 1, 0])
edges = BCOO((values, indices), shape=(4, 2)).sum_duplicates(remove_zeros=False).sort_indices()

print(*[(edge.indices[:, 0], edge.data) for edge in edges], sep="\n")
```
Result:

```
(Array([0, 1], dtype=int32), Array([0, 0], dtype=int32))
(Array([0, 2], dtype=int32), Array([0, 0], dtype=int32))
(Array([1, 2], dtype=int32), Array([0, 0], dtype=int32))
(Array([0, 1], dtype=int32), Array([1, 0], dtype=int32))
```

The second and third row both contain the index 2. This should not be.

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.30
jaxlib: 0.4.30
numpy:  1.26.4
python: 3.9.20 (main, Sep  7 2024, 18:35:26)  [GCC 9.4.0]
jax.devices (1 total, 1 local): [cuda(id=0)]
process_count: 1
platform: uname_result(system='Linux', node='tommotower', release='5.15.0-122-generic', version='#132~20.04.1-Ubuntu SMP Fri Aug 30 15:50:07 UTC 2024', machine='x86_64')


$ nvidia-smi
Sat Oct  5 12:15:18 2024       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 555.42.06              Driver Version: 555.42.06      CUDA Version: 12.5     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce RTX 4070 Ti     Off |   00000000:01:00.0  On |                  N/A |
|  0%   42C    P0             43W /  285W |    9643MiB /  12282MiB |      2%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+------------------------+----------------------+
                                                                                         
+-----------------------------------------------------------------------------------------+
| Processes:                                                                              |
|  GPU   GI   CI        PID   Type   Process name                              GPU Memory |
|        ID   ID                                                               Usage      |
|=========================================================================================|
|    0   N/A  N/A      1482      G   /usr/lib/xorg/Xorg                            247MiB |
|    0   N/A  N/A      1635      G   /usr/bin/gnome-shell                          191MiB |
|    0   N/A  N/A     11111      G   ...11,262144 --variations-seed-version         43MiB |
|    0   N/A  N/A     28946      C   ...envs/probabilistic_model/bin/python       9132MiB |
+-----------------------------------------------------------------------------------------+
```

**Comment by jakevdp** [comment]
Thanks for the question! This is working as expected: BCOO uses out-of-bound indices in its representation to mark padded entries. When you iterate over a sparse array, you get a sequence of outputs with the same `nse`. Since some of your rows have two elements defined and some have only one, the rows with one element need this padding.

**Comment by tomsch420** [comment]
Thanks for the quick reply! I see, are there more situations where one needs to know such things?

For the curious, here is an innerloop begin that removes the out-of-bounds elements:

```
edge_indices = edges.indices[:, 0]
in_bound_elements = edge_indices < edges.nse
edge_indices = edge_indices[in_bound_elements]
edge_data = edges.data[in_bound_elements]

```

**Comment by jakevdp** [comment]
You can also remove out-of-bound indices from a BCOO array `x` using `x.sum_duplicates()` or `x.sum_duplicates(remove_zeros=False)`, depending on whether you want to keep the explicit zeros in the representation.

**Comment by tomsch420** [comment]
Thanks, with the `x.sum_duplicates(remove_zeros=False)` I could replace some parts of the code and it got cleaner!
